File: dispotrains.webapp/src/storage/station.go
package storage

import (
	"strings"
	"time"
)

type Coordinates struct {
	Latitude  float64
	Longitude float64
}

type Station struct {
	DisplayName  string
	Name         string
	City         string
	Position     Coordinates
	OsmID        string
	Lines        []*Line
	Elevators    []*Elevator
	Code         string
	HasElevators bool
	LastUpdate   time.Time
}

func NewStation(name, city, code string) *Station {
	var station *Station = new(Station)
	station.Name = strings.Replace(strings.TrimSpace(name), "/", "-", -1)
	station.DisplayName = computeDisplayName(station.Name)
	station.City = strings.TrimSpace(city)
	station.Code = code
	station.HasElevators = true
	return station
}

func computeDisplayName(name string) string {
	var displayName string = strings.Title(strings.ToLower(name))
	if displayName != "Gare De Lyon" && displayName != "Gare Du Nord" {
		displayName = strings.Replace(displayName, "Gare De ", "", 1)
		displayName = strings.Replace(displayName, "Gare Du ", "", 1)
	}
	displayName = strings.Replace(displayName, " De ", " de ", -1)
	displayName = strings.Replace(displayName, " Du ", " du ", -1)
	displayName = strings.Replace(displayName, " D ", " d'", -1)
	displayName = strings.Replace(displayName, " D'", " d'", -1)
	displayName = strings.Replace(displayName, " Le ", " le ", -1)
	displayName = strings.Replace(displayName, " Les ", " les ", -1)
	displayName = strings.Replace(displayName, " La ", " la ", -1)
	displayName = strings.Replace(displayName, " L ", " l'", -1)
	displayName = strings.Replace(displayName, " En ", " en ", -1)
	displayName = strings.Replace(displayName, " Au ", " au ", -1)
	displayName = strings.Replace(displayName, " Aux ", " aux ", -1)
	displayName = strings.Replace(displayName, " A ", " à ", -1)
	displayName = strings.Replace(displayName, " Sur ", " sur ", -1)
	displayName = strings.Replace(displayName, " Sous ", " sous ", -1)
	return displayName
}

func NewRampStation(name, city string) *Station {
	var station *Station = new(Station)
	station.Name = strings.Replace(strings.TrimSpace(name), "/", "-", -1)
	station.DisplayName = computeDisplayName(station.Name)
	station.City = strings.TrimSpace(city)
	station.HasElevators = false
	return station
}

func (station *Station) AttachLine(line *Line) {
	station.Lines = append(station.Lines, line)
	line.attachStation(station)
}

func (station *Station) NewElevator(id, situation, direction string) *Elevator {
	var elevator *Elevator = new(Elevator)
	elevator.ID = strings.TrimSpace(id)
	elevator.Situation = strings.TrimSpace(situation)
	elevator.Direction = strings.TrimSpace(direction)
	elevator.station = station
	station.Elevators = append(station.Elevators, elevator)
	return elevator
}

func (station *Station) GetElevators() []*Elevator {
	return station.Elevators
}

func (station *Station) Available() bool {
	for _, elevator := range station.Elevators {
		status := elevator.GetLastStatus()
		if status != nil && status.State != "Disponible" {
			return false
		}
	}
	return true
}

File: dispotrains.webapp/src/web/src/app/station.ts
export class NetworkStats {
  good: number;
  bad: number;
  longBad: number;

  constructor(data: Object) {
    this.good = data['Good'];
    this.bad = data['Bad'];
    this.longBad = data['LongBad'];
  }

  Total(): number {
    return this.good + this.bad;
  }

  PercentBad(): number {
    return (this.bad * 100) / (this.Total());
  }

  PercentLongBad(): number {
    return (this.longBad * 100) / (this.Total());
  }
}

export class Line {
  network: string;
  id: string;

  constructor(lineData: Object) {
    this.network = lineData['network'];
    this.id = lineData['id'];
  }

  GetName(): string {
    return this.network + ' ' + this.id;
  }
}

export class Position {
  latitude: number;
  longitude: number;

  constructor(positionData: Object) {
    this.latitude = positionData['latitude'];
    this.longitude = positionData['longitude'];
  }
}

export class Status {
  lastupdate: string;
  state: string;
  forecast: string;

  constructor(statusData: Object) {
    if (statusData !== undefined && statusData !== null) {
      this.lastupdate = statusData['lastupdate'];
      this.state = statusData['state'];
      this.forecast = statusData['forecast'];
    } else {
      this.lastupdate = '';
      this.state = 'Information non disponible';
      this.forecast = undefined;
    }
  }
}

export class Elevator {
  direction: string;
  id: string;
  situation: string;
  status: Status;

  constructor(elevatorData: Object) {
    this.direction = elevatorData['direction'];
    this.id = elevatorData['id'];
    this.situation = elevatorData['situation'];
    this.status = new Status(elevatorData['status']);
  }

  available(): boolean {
    return this.status.state === 'Disponible';
  }
}

export class Station {
  lines: Line[];
  elevators: Elevator[];
  displayname: string;
  name: string;
  city: string;
  position: Position;
  osmid: string;

  constructor(stationData: Object) {
    this.name = stationData['name'];
    this.displayname = stationData['displayname'];
    this.city = stationData['city'];
    this.osmid = stationData['osmid'];
    if (stationData['position'] !== undefined) {
      this.position = new Position(stationData['position']);
    }

    this.lines = new Array<Line>();
    for (let lineData of stationData['lines']) {
      this.lines.push(new Line(lineData));
    }

    this.elevators = new Array<Elevator>();
    for (let elevatorData of stationData['elevators']) {
      this.elevators.push(new Elevator(elevatorData));
    }
  }

  available(): boolean {
    for (let elevator of this.elevators) {
      if (!elevator.available()) {
        return false;
      }
    }
    return true;
  }
}
